File: android/app/src/main/java/com/teykaijun/adblocker/dns/DomainMatcher.kt
package com.teykaijun.adblocker.dns

/**
 * Decides whether a hostname is blocked. A rule for `example.com` also covers
 * every subdomain, and an allow rule anywhere up the chain wins over a block
 * rule, the same as `@@||example.com^` in Adblock Plus syntax.
 */
class DomainMatcher(private val blocked: Set<String>, private val allowed: Set<String>) {

    val blockedCount: Int get() = blocked.size

    fun isBlocked(domain: String): Boolean {
        var host = domain.trimEnd('.').lowercase()
        var blockHit = false
        while (host.isNotEmpty()) {
            if (host in allowed) return false
            if (!blockHit && host in blocked) blockHit = true
            val dot = host.indexOf('.')
            if (dot < 0) break
            host = host.substring(dot + 1)
        }
        return blockHit
    }

    companion object {
        val EMPTY = DomainMatcher(emptySet(), emptySet())
    }
}

/**
 * What the VPN answers with: every rule in [matcher], and the subset from the scam lists in
 * [scam], so a blocked scam site can be reported instead of just failing to load.
 */
data class Rules(val matcher: DomainMatcher, val scam: DomainMatcher) {
    fun isScam(domain: String) = scam.isBlocked(domain) && matcher.isBlocked(domain)

    companion object {
        val EMPTY = Rules(DomainMatcher.EMPTY, DomainMatcher.EMPTY)
    }
}

File: windows/src/AdBlocker/Filtering/DomainMatcher.cs
namespace AdBlocker.Filtering;

public readonly record struct MatchResult(bool Blocked, string? Rule);

/// <summary>
/// Decides whether a hostname is blocked. A rule for <c>example.com</c> also covers
/// every subdomain, and an allow rule anywhere up the chain wins over a block rule,
/// the same as <c>@@||example.com^</c> in Adblock Plus syntax.
/// </summary>
public sealed class DomainMatcher
{
    public static DomainMatcher Empty { get; } = new([], []);

    private readonly HashSet<string>.AlternateLookup<ReadOnlySpan<char>> _blocked;
    private readonly HashSet<string>.AlternateLookup<ReadOnlySpan<char>> _allowed;

    public DomainMatcher(IEnumerable<string> blocked, IEnumerable<string> allowed)
    {
        var blockedSet = blocked as HashSet<string> is { } b && b.Comparer == StringComparer.Ordinal ? b : new HashSet<string>(blocked, StringComparer.Ordinal);
        var allowedSet = allowed as HashSet<string> is { } a && a.Comparer == StringComparer.Ordinal ? a : new HashSet<string>(allowed, StringComparer.Ordinal);
        _blocked = blockedSet.GetAlternateLookup<ReadOnlySpan<char>>();
        _allowed = allowedSet.GetAlternateLookup<ReadOnlySpan<char>>();
        BlockedCount = blockedSet.Count;
    }

    public int BlockedCount { get; }

    /// <param name="name">A lower-case name without a trailing dot, as <see cref="Dns.DnsMessage"/> produces.</param>
    public bool IsBlocked(ReadOnlySpan<char> name)
    {
        var hit = false;
        for (var host = name; !host.IsEmpty;)
        {
            if (_allowed.Contains(host)) return false;
            hit = hit || _blocked.Contains(host);
            var dot = host.IndexOf('.');
            if (dot < 0) break;
            host = host[(dot + 1)..];
        }
        return hit;
    }

    /// <summary>Like <see cref="IsBlocked"/>, but also reports which rule decided.</summary>
    public MatchResult Match(string domain)
    {
        string? blockedBy = null;
        for (var host = domain.TrimEnd('.').ToLowerInvariant().AsSpan(); !host.IsEmpty;)
        {
            if (_allowed.Contains(host)) return new MatchResult(false, host.ToString());
            if (blockedBy is null && _blocked.Contains(host)) blockedBy = host.ToString();
            var dot = host.IndexOf('.');
            if (dot < 0) break;
            host = host[(dot + 1)..];
        }
        return new MatchResult(blockedBy is not null, blockedBy);
    }
}
